common/persistence/sql/sqlplugin/sqlite/conn_pool.go

```diff
@@ -0,0 +1,104 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package sqlite
+
+import (
+	"github.com/jmoiron/sqlx"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/resolver"
+	"sync"
+)
+
+type connPool struct {
+	mu   sync.Mutex
+	pool map[string]entry
+}
+
+type entry struct {
+	db       *sqlx.DB
+	refCount int
+}
+
+func newConnPool() *connPool {
```

Comment:
connection pool exists here due to frontend / matching / history / worker (4 services) each will create its own DB connection & causing DB locking issue?

can you plz put a comment here?
we can possibly improve the DI to initialize this DB connection only once (for single binary), then this conn pool management can possibly be removed

Reply:
Comment added. Yes, refactoring DI would be the ideal approach, but I'm not sure about the size of the impact.

Reply:
Can we also try to turn off cluster, gossip and other services irrelevant for single binary server?

Reply:
Also, some services open and close connections sequentially, getting DB connection stopped and destroying the database in memory.

Reply:
>Can we also try to turn off cluster, gossip and other services irrelevant for single binary server?

we definitely should (assuming we are talking about the ideal version), plus inter-service communication can by pass network layer.

>Also, some services open and close connections sequentially, getting DB connection stopped and destroying the database in memory.

sorry what do you mean? if using DI, DI should only initialize the DB conn once (assuming 1 and only 1 conn)?

Reply:
This use case won't exist with DI, but it exists now. We can probably take a look at it once it will be time to clean up all unnecessary services.